File: src/2025-02-24-deepclaude-rust/go/exercises/practice/poker/poker.go
package poker

import (
	"fmt"
	"sort"
	"strings"
)

type Card struct {
	Rank  string
	Suit  string
	Value int
}

func BestHand(hands []string) ([]string, error) {
	if len(hands) == 0 {
		return nil, nil
	}

	var bestHands []string
	var maxScore []int

	for _, handStr := range hands {
		score, err := evaluateHand(handStr)
		if err != nil {
			return nil, err
		}

		if len(bestHands) == 0 {
			bestHands = []string{handStr}
			maxScore = score
			continue
		}

		cmp := compareScores(score, maxScore)
		if cmp > 0 {
			bestHands = []string{handStr}
			maxScore = score
		} else if cmp == 0 {
			bestHands = append(bestHands, handStr)
		}
	}

	return bestHands, nil
}

func evaluateHand(handStr string) ([]int, error) {
	cards, err := parseHand(handStr)
	if err != nil {
		return nil, err
	}

	values := make([]int, len(cards))
	suits := make([]string, len(cards))
	for i, card := range cards {
		values[i] = card.Value
		suits[i] = card.Suit
	}

	flush := isFlush(suits)
	straight, high := isStraight(values)

	if flush && straight {
		return []int{8, high}, nil
	}

	counts := countRanks(values)
	pairs := make([]int, 0)
	three := 0
	four := 0

	for rank, count := range counts {
		switch count {
		case 2:
			pairs = append(pairs, rank)
		case 3:
			three = rank
		case 4:
			four = rank
		}
	}

	if four > 0 {
		kickers := make([]int, 0)
		for _, v := range values {
			if v != four {
				kickers = append(kickers, v)
			}
		}
		return []int{7, four, kickers[0]}, nil
	}

	if three > 0 && len(pairs) >= 1 {
		return []int{6, three, pairs[0]}, nil
	}

	if flush {
		sorted := make([]int, len(values))
		copy(sorted, values)
		sort.Sort(sort.Reverse(sort.IntSlice(sorted)))
		return append([]int{5}, sorted...), nil
	}

	if straight {
		return []int{4, high}, nil
	}

	if three > 0 {
		kickers := make([]int, 0)
		for _, v := range values {
			if v != three {
				kickers = append(kickers, v)
			}
		}
		sort.Sort(sort.Reverse(sort.IntSlice(kickers)))
		return append([]int{3, three}, kickers...), nil
	}

	if len(pairs) >= 2 {
		sort.Sort(sort.Reverse(sort.IntSlice(pairs)))
		kicker := 0
		for _, v := range values {
			if v != pairs[0] && v != pairs[1] {
				kicker = v
				break
			}
		}
		return []int{2, pairs[0], pairs[1], kicker}, nil
	}

	if len(pairs) == 1 {
		kickers := make([]int, 0)
		for _, v := range values {
			if v != pairs[0] {
				kickers = append(kickers, v)
			}
		}
		sort.Sort(sort.Reverse(sort.IntSlice(kickers)))
		return append([]int{1, pairs[0]}, kickers...), nil
	}

	sorted := make([]int, len(values))
	copy(sorted, values)
	sort.Sort(sort.Reverse(sort.IntSlice(sorted)))
	return append([]int{0}, sorted...), nil
}

func parseHand(handStr string) ([]Card, error) {
	cardStrs := strings.Split(handStr, " ")
	if len(cardStrs) != 5 {
		return nil, fmt.Errorf("invalid hand: %s", handStr)
	}

	cards := make([]Card, 0, 5)
	for _, cs := range cardStrs {
		card, err := parseCard(cs)
		if err != nil {
			return nil, err
		}
		cards = append(cards, card)
	}

	sort.Slice(cards, func(i, j int) bool {
		return cards[i].Value < cards[j].Value
	})

	return cards, nil
}

func parseCard(cardStr string) (Card, error) {
	if len(cardStr) < 2 {
		return Card{}, fmt.Errorf("invalid card: %s", cardStr)
	}

	runes := []rune(cardStr)
	suitRune := runes[len(runes)-1]
	suit := string(suitRune)
	rankRunes := runes[:len(runes)-1]
	rankStr := string(rankRunes)

	value, ok := rankValues[rankStr]
	if !ok {
		return Card{}, fmt.Errorf("invalid rank: %s", rankStr)
	}

	return Card{
		Rank:  rankStr,
		Suit:  suit,
		Value: value,
	}, nil
}

var rankValues = map[string]int{
	"2":  2,
	"3":  3,
	"4":  4,
	"5":  5,
	"6":  6,
	"7":  7,
	"8":  8,
	"9":  9,
	"10": 10,
	"J":  11,
	"Q":  12,
	"K":  13,
	"A":  14,
}

func isFlush(suits []string) bool {
	first := suits[0]
	for _, s := range suits[1:] {
		if s != first {
			return false
		}
	}
	return true
}

func isStraight(values []int) (bool, int) {
	sorted := make([]int, len(values))
	copy(sorted, values)
	sort.Ints(sorted)

	// Check normal straight
	isConsecutive := true
	for i := 1; i < len(sorted); i++ {
		if sorted[i] != sorted[i-1]+1 {
			isConsecutive = false
			break
		}
	}

	if isConsecutive {
		return true, sorted[len(sorted)-1]
	}

	// Check Ace-low straight (A,2,3,4,5)
	if sorted[0] == 2 && sorted[1] == 3 && sorted[2] == 4 && sorted[3] == 5 && sorted[4] == 14 {
		return true, 5
	}

	return false, 0
}

func countRanks(values []int) map[int]int {
	counts := make(map[int]int)
	for _, v := range values {
		counts[v]++
	}
	return counts
}

func compareScores(a, b []int) int {
	for i := 0; i < len(a) && i < len(b); i++ {
		if a[i] > b[i] {
			return 1
		} else if a[i] < b[i] {
			return -1
		}
	}
	if len(a) > len(b) {
		return 1
	} else if len(a) < len(b) {
		return -1
	}
	return 0
}

File: src/2025-02-24-deepclaude-rust/javascript/exercises/practice/poker/poker.js
//
// This is only a SKELETON file for the 'Poker' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const bestHands = (hands) => {
  if (hands.length === 0) return [];

  const evaluateHand = (hand) => {
    const cards = hand.split(' ');
    const rankOrder = { '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8, '9': 9, 'T': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14 };
    const suits = [];
    const ranks = [];
    for (const card of cards) {
      let rankChar;
      let suit;
      if (card.startsWith('10')) {
        rankChar = 'T';
        suit = card[2];
      } else {
        rankChar = card[0];
        suit = card[1];
      }
      suits.push(suit);
      ranks.push(rankOrder[rankChar]);
    }
    ranks.sort((a, b) => a - b);

    const isFlush = suits.every(s => s === suits[0]);
    const isStraight = (() => {
      if (ranks[4] - ranks[0] === 4 && new Set(ranks).size === 5) return true;
      if (ranks[0] === 2 && ranks[1] === 3 && ranks[2] === 4 && ranks[3] === 5 && ranks[4] === 14) return true;
      return false;
    })();

    const counts = {};
    for (const rank of ranks) {
      counts[rank] = (counts[rank] || 0) + 1;
    }
    const countValues = Object.values(counts).sort((a, b) => b - a);

    let rank, values = [];
    if (isStraight && isFlush) {
      rank = 8;
      const highCard = (ranks[0] === 2 && ranks[4] === 14) ? 5 : ranks[4];
      values = [highCard];
    } else if (countValues[0] === 4) {
      rank = 7;
      const quadRank = parseInt(Object.keys(counts).find(k => counts[k] === 4));
      const kicker = parseInt(Object.keys(counts).find(k => counts[k] === 1));
      values = [quadRank, kicker];
    } else if (countValues[0] === 3 && countValues[1] === 2) {
      rank = 6;
      const triplet = parseInt(Object.keys(counts).find(k => counts[k] === 3));
      const pair = parseInt(Object.keys(counts).find(k => counts[k] === 2));
      values = [triplet, pair];
    } else if (isFlush) {
      rank = 5;
      values = [...ranks].sort((a, b) => b - a);
    } else if (isStraight) {
      rank = 4;
      const highCard = (ranks[0] === 2 && ranks[4] === 14) ? 5 : ranks[4];
      values = [highCard];
    } else if (countValues[0] === 3) {
      rank = 3;
      const triplet = parseInt(Object.keys(counts).find(k => counts[k] === 3));
      const kickers = ranks.filter(r => r !== triplet).sort((a, b) => b - a);
      values = [triplet, ...kickers];
    } else if (countValues[0] === 2 && countValues[1] === 2) {
      rank = 2;
      const pairs = Object.entries(counts)
        .filter(([k, v]) => v === 2)
        .map(([k, v]) => parseInt(k))
        .sort((a, b) => b - a);
      const kicker = parseInt(Object.entries(counts).find(([k, v]) => v === 1)[0]);
      values = [...pairs, kicker];
    } else if (countValues[0] === 2) {
      rank = 1;
      const pairRank = parseInt(Object.keys(counts).find(k => counts[k] === 2));
      const kickers = ranks.filter(r => r !== pairRank).sort((a, b) => b - a);
      values = [pairRank, ...kickers];
    } else {
      rank = 0;
      values = [...ranks].sort((a, b) => b - a);
    }

    return { hand, key: [rank, ...values] };
  };

  const compareKeys = (a, b) => {
    for (let i = 0; i < Math.max(a.length, b.length); i++) {
      const aVal = a[i] || 0;
      const bVal = b[i] || 0;
      if (aVal > bVal) return 1;
      if (aVal < bVal) return -1;
    }
    return 0;
  };

  const evaluated = hands.map(evaluateHand);
  let maxKey = evaluated[0].key;
  evaluated.forEach(ev => {
    if (compareKeys(ev.key, maxKey) > 0) {
      maxKey = ev.key;
    }
  });

  const best = evaluated.filter(ev => compareKeys(ev.key, maxKey) === 0).map(ev => ev.hand);
  return best;
};
